### ISSUE TITLE
Single class object detection with extreme number of null images

### ISSUE THREAD

**Issue by fabi34an** [ultralytics/ultralytics#22523]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hey guys,

i want to perform a single class object detection on my custom dataset (detect fractures of one specific bone in ultrasound pictures) using YOLOv11 (probably medium size). My dataset is quite small regarding patients (20 healthy, 20 fractured), but per patient I have 10 recordings containing each around 700 images. The problem is that only a small percentage shows the relevant bone, so the amount of images that contain an object to be detected is very small. In numbers: 273260 images, 269110 backgrounds (I think also referred to as null images?). 
The model is also supposed to learn that there is no fracture present in the image, so I want to include those pictures in the training too (empty label .txt). If I have time I will address the black/empty images in the beginning and end of a recording sequence (around 100 per recording), but right now I need to focus on getting some results. Also for us it's not important to have an exactly overlapping bounding box, but to detect the fracture at all. 
I am really helpful for recommendations addressing:

- the class imbalance: I know of loss weighting (tuning parameters ,,box" and ,,pls"), but I am afraid the model will just learn that ,,no fracture" is the easiest answer
- making the most of my data: I am thinking about performing a 5Fold CV (with 80% train, 20% test - no validation unfortunately) or a 5Fold CV with a holdout test set (90% for train and val, 10% for final inference on test subset)
-general recommendations for my case

I am grateful for answers!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @fabi34an, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will also follow up with you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (e.g., single-class training on medical ultrasound with many empty images using YOLO11), please provide as much information as possible, including:
- dataset image examples (showing positive and empty frames)
- a brief summary of your split strategy (e.g., folds, patient-level separation)
- your current model/hyperparameters and training logs
- confirmation you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)
This context helps us give targeted guidance faster 🙌

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You shouldn't include all of them at first. Only make 20-30% of the images as background. Then train a model. Then run the model on all your background images. And then add only images where there was a detection since they are false positives as background. And retrain.

**Comment by fabi34an** [comment]
@Y-T-G thank you for the reply. that sounds interesting, can you explain why and how that affects the training? is that a typical approach?

**Comment by Y-T-G** [comment]
Too many background images would cause problems in training and also lower recall. You should only add background images that actually have false positives. Otherwise, you're adding images that the model already doesn't have a problem avoiding false positives in. By running the trained model on your background images and only selecting the ones with false positives, you're only using background images that the model struggles with

**Comment by fabi34an** [comment]
so is there no such thing as ,,the model learns where not to detect something/what a healthy structure looks like" ? our idea is to later use the trained model in a real-time application where it will definitely see all the null/background images. that's where the idea of training it on them comes from.

**Comment by Y-T-G** [comment]
That's what background images are for. But you're not supposed to have that as 98% of your data. The model will not learn anything from positive images for most of the epoch if you do that. It's also redundant when you add images that the model doesn't struggle with to begin with.

**Comment by fabi34an** [comment]
thanks a lot @Y-T-G ! 
@glenn-jocher can you add any helpful information from your side? besides the described approach by Y-T-G and tuning the cls parameter?

**Comment by glenn-jocher** [comment]
Good question—beyond Y‑T‑G’s hard‑negative mining, I’d do three things: 1) keep all positives and cap backgrounds to ~1–3× positives per epoch by oversampling positives via a `train.txt` (list positive image paths multiple times), 2) mine only the backgrounds that produce false positives with your first model (example below), and 3) split by patient (not frame) with a small validation set per fold; a 5‑fold CV plus a held‑out patient test set is a solid plan for unbiased recall estimates.

```python
# Hard-negative mining: copy only background frames that trigger detections
from ultralytics import YOLO
from pathlib import Path
import shutil

model = YOLO('runs/detect/train/weights/best.pt')
bg_dir = 'datasets/mydata/backgrounds'          # directory of empty images
dst = Path('datasets/mydata/neg_mined'); dst.mkdir(parents=True, exist_ok=True)

for r in model.predict(source=bg_dir, conf=0.1, stream=True, imgsz=640):
    if len(r.boxes):  # false positive present
        shutil.copy(r.path, dst / Path(r.path).name)
```

Because you don’t need box precision, prioritize recall: keep coarse boxes, train normally, and tune your inference `conf` threshold down for deployment. If you want to experiment with loss shaping for imbalance, consider focal loss for classification (see the Ultralytics [Focal Loss](https://www.ultralytics.com/glossary/focal-loss) overview) and keep augmentation ultrasound‑appropriate using the [Albumentations integration](https://docs.ultralytics.com/integrations/albumentations/). For general pitfalls around class imbalance and splits, the notes in [Troubleshooting common YOLO issues](https://docs.ultralytics.com/guides/yolo-common-issues/) may help.

**Comment by fabi34an** [comment]
thanks for the helpful reply @glenn-jocher 
I still have some uncertainties:

1.  does it help if the model sees the same images multiple times? I had read before it doesn't 

2. how can I put more focus on recall? like is there any way to emphasize certain metrics during training/validation?

**Comment by glenn-jocher** [comment]
Great follow‑ups.

1) Yes—under extreme imbalance, repeating positives per epoch helps because YOLO applies on‑the‑fly augmentation, so each pass isn’t identical; prefer oversampling positives (via a `train.txt` that lists positive image paths multiple times) rather than flooding training with nulls.

2) To favor recall: modestly up‑weight classification loss and select checkpoints by validation recall, then deploy with a lower `conf`. Example:
```bash
# Train with slightly higher cls loss, keep all checkpoints
yolo train model=yolo11m.pt data=mydata.yaml epochs=100 hyp=hyp_recall.yaml save_period=1

# Evaluate with a low confidence to bias toward recall; choose the epoch with highest recall from results.csv
yolo val model=runs/detect/train/weights/best.pt data=mydata.yaml conf=0.05 iou=0.7
```
```yaml
# hyp_recall.yaml
cls: 1.5
```
For robust patient‑level estimates and threshold calibration, see the guide on [K‑Fold Cross‑Validation with Ultralytics](https://docs.ultralytics.com/guides/kfold-cross-validation/); for what recall measures and the precision‑recall trade‑off, see our [Recall overview](https://www.ultralytics.com/glossary/recall).

**Comment by fabi34an** [comment]
thanks a lot!
I will try those approaches and see what I can get out of my data.

Actually I am thinking about changing my class structure. For object detection in medical imaging do you generally advise to train the model with 2 classes (healthy structure, structure with anomaly)? we had the focus on detecting the fracture so we started with 1 class (bounding box around the fracture). now I am doubting if the model learns like that what we want... does the model just learn what's inside the bounding box ground truth or does it understand the bigger picture? a bounding box around the whole bone (that is partially broken at one place) is not really what we were looking for, but maybe what the model needs? if it's the case, would the 2nd class be a bounding box around the whole bone again or (like we are doing now) around just the crack?

a bit of a topic change away from the technical part more to the usage of yolo, but I very much appreciate your help @glenn-jocher and other people!!

**Comment by glenn-jocher** [comment]
Great question—choose labels that match the decision you need. If your goal is simply “flag any fracture,” keep a single class `fracture` and box the crack itself (slightly oversized is fine); treat healthy frames as background. Use two classes only if you want ROI-level bone health (“bone_healthy” vs “bone_fractured”), in which case draw the same bone ROI in every frame and label it healthy or fractured consistently—mixing crack-only boxes for one class and whole-bone boxes for the other will confuse training. YOLO does use surrounding context, but supervision comes from what’s inside each box, so avoid whole-bone boxes for a one‑class crack detector. For when to prefer anomaly-style vs explicit class labels, see our anomaly‑detection guidance in [Vision Anomaly Detection with YOLO11](https://www.ultralytics.com/blog/vision-ai-for-anomaly-detection-a-quick-overview), and for tips on consistent annotation, see [Data Collection and Annotation Strategies](https://docs.ultralytics.com/guides/data-collection-and-annotation/).

**Comment by fabi34an** [comment]
how would I actually implement the use of focal loss? 

I found this:

from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.utils.loss import FocalLoss

class CustomDetectionTrainer(DetectionTrainer):
    def get_model(self, cfg, weights):
        model = super().get_model(cfg, weights)
        # Now we can properly pass parameters during initialization
        model.model[-1].loss_fn = FocalLoss(gamma=2.0, alpha=0.25)
        return model

**Comment by glenn-jocher** [comment]
Nice find, but for YOLO11 you don’t want to overwrite `model.model[-1].loss_fn`—the detection loss is handled by `v8DetectionLoss` inside the model, not by a `loss_fn` on the last layer, so that snippet won’t actually switch the training loss.

The supported way to use focal loss for detection is to enable it via the training hyperparameter and let `v8DetectionLoss` wire `FocalLoss` in for the classification branch internally. In Python:

```python
from ultralytics import YOLO

model = YOLO("yolo11m.pt")
model.train(data="mydata.yaml", fl_gamma=1.5)  # try 1.0–2.0 range
```

This keeps the standard IoU/DFL box loss and swaps the classification term to focal loss with your chosen `gamma`. If you ever need lower‑level control, the loss building blocks (including `FocalLoss`) are documented in the [Ultralytics loss reference](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.FocalLoss).
